<a href="https://colab.research.google.com/github/liznjoki/liznjoki/blob/main/Suicide_Rates_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Suicide Rates Overview (1985 to 2021)
#### Date Source is Kaggle.
#### The data contains information on country, year, sex, age group, count of suicides, population, suicide rate, country-year composite key, HDI for year, GDP for year, GDP per capita, and generation (based on age grouping average).
####Information is useful in suicide prevention.


In [ ]:
pip install opendatasets --upgrade --quiet

#Dowload the Data

In [ ]:

pip install pycountry-convert

In [ ]:
import opendatasets as od
download_url = "https://www.kaggle.com/datasets/omkargowda/suicide-rates-overview-1985-to-2021"
od.download(download_url )

In [ ]:
data_filename= '/content/suicide-rates-overview-1985-to-2021/master.csv'

#Data Preparation and Cleaning
1. Load the file using Pandas
2. Look at some information about the data and columns
3.Fix any incorrect or missing values

In [ ]:
import pandas as pd
df= pd.read_csv(data_filename)
df.head(10)

country  year     sex          age  suicides_no  population  \
0  Albania  1987    male  15-24 years         21.0      312900   
1  Albania  1987    male  35-54 years         16.0      308000   
2  Albania  1987  female  15-24 years         14.0      289700   
3  Albania  1987    male    75+ years          1.0       21800   
4  Albania  1987    male  25-34 years          9.0      274300   
5  Albania  1987  female    75+ years          1.0       35600   
6  Albania  1987  female  35-54 years          6.0      278800   
7  Albania  1987  female  25-34 years          4.0      257200   
8  Albania  1987    male  55-74 years          1.0      137500   
9  Albania  1987  female   5-14 years          0.0      311000   

   suicides/100k pop country-year  HDI for year  gdp_for_year ($)   \
0               6.71  Albania1987           NaN     2,15,66,24,900   
1               5.19  Albania1987           NaN     2,15,66,24,900   
2               4.83  Albania1987           NaN     2,15,66,24,900   
3               4.59  Albania1987           NaN     2,15,66,24,900   
4               3.28  Albania1987           NaN     2,15,66,24,900   
5               2.81  Albania1987           NaN     2,15,66,24,900   
6               2.15  Albania1987           NaN     2,15,66,24,900   
7               1.56  Albania1987           NaN     2,15,66,24,900   
8               0.73  Albania1987           NaN     2,15,66,24,900   
9               0.00  Albania1987           NaN     2,15,66,24,900   

   gdp_per_capita ($)       generation  
0               796.0     Generation X  
1               796.0           Silent  
2               796.0     Generation X  
3               796.0  G.I. Generation  
4               796.0          Boomers  
5               796.0  G.I. Generation  
6               796.0           Silent  
7               796.0          Boomers  
8               796.0  G.I. Generation  
9               796.0     Generation X

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31756 entries, 0 to 31755
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             31756 non-null  object 
 1   year                31756 non-null  int64  
 2   sex                 31756 non-null  object 
 3   age                 31756 non-null  object 
 4   suicides_no         30556 non-null  float64
 5   population          31756 non-null  int64  
 6   suicides/100k pop   31756 non-null  float64
 7   country-year        31756 non-null  object 
 8   HDI for year        12300 non-null  float64
 9    gdp_for_year ($)   31756 non-null  object 
 10  gdp_per_capita ($)  31756 non-null  float64
 11  generation          31756 non-null  object 
dtypes: float64(4), int64(2), object(6)
memory usage: 2.9+ MB


In [ ]:
df.rename(columns=lambda x: x.lstrip(), inplace=True)
df.rename(columns=lambda x: x.replace(" ", "_").replace("-", "_"), inplace=True)

In [ ]:
df.describe()

year   suicides_no    population  suicides/100k_pop  \
count  31756.000000  30556.000000  3.175600e+04       31756.000000   
mean    2003.395390    237.136242  7.217454e+06          11.716668   
std        9.760604    867.956107  5.799323e+07          21.590926   
min     1985.000000      0.000000  2.780000e+02           0.000000   
25%     1996.000000      3.000000  1.288210e+05           0.370000   
50%     2004.000000     25.000000  5.468325e+05           4.285000   
75%     2011.000000    132.000000  2.908763e+06          14.560000   
max     2020.000000  22338.000000  1.411100e+09         515.092734   

       HDI_for_year  gdp_per_capita_($)  
count  12300.000000        31756.000000  
mean       0.793666        17588.954606  
std        0.097480        19464.861140  
min        0.377790          251.000000  
25%        0.727000         3765.000000  
50%        0.800457        10062.000000  
75%        0.874000        25622.000000  
max        0.975114       126352.000000

In [ ]:
numerics = ['int16', 'int32', 'int64', 'float64']
numeric_columns = df.select_dtypes(include=numerics)
numeric_columns.columns

In [ ]:
import matplotlib.pyplot as plt
percentage_of_null_values =df.isnull().sum()/len(df)*100
percentage_of_null_values

In [ ]:
(df.isnull().sum()/len(df)).sort_values(ascending=False)

HDI_for_year          0.612672
suicides_no           0.037788
country               0.000000
year                  0.000000
sex                   0.000000
age                   0.000000
population            0.000000
suicides/100k_pop     0.000000
country_year          0.000000
gdp_for_year_($)_     0.000000
gdp_per_capita_($)    0.000000
generation            0.000000
dtype: float64

In [ ]:
df.columns

Index(['country', 'year', 'sex', 'age', 'suicides_no', 'population',
       'suicides/100k_pop', 'country_year', 'HDI_for_year',
       'gdp_for_year_($)_', 'gdp_per_capita_($)', 'generation'],
      dtype='object')

In [ ]:
df.dropna(subset= ["suicides_no"], inplace=True)

#Exploratory Data Analysis
Analyse:
1. Country
2. Continent
3. GDP for year
4. Generation
5. Population
6. Age

###Country

In [ ]:
import matplotlib.pyplot as plt
df.country.unique()

array(['Albania', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Barbados', 'Belarus', 'Belgium', 'Belize',
       'Bosnia and Herzegovina', 'Brazil', 'Bulgaria', 'Cabo Verde',
       'Canada', 'Chile', 'Colombia', 'Costa Rica', 'Croatia', 'Cuba',
       'Cyprus', 'Czech Republic', 'Denmark', 'Dominica', 'Ecuador',
       'El Salvador', 'Estonia', 'Fiji', 'Finland', 'France', 'Georgia',
       'Germany', 'Greece', 'Grenada', 'Guatemala', 'Guyana', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Jamaica', 'Japan',
       'Kazakhstan', 'Kiribati', 'Kuwait', 'Kyrgyzstan', 'Latvia',
       'Lithuania', 'Luxembourg', 'Macau', 'Maldives', 'Malta',
       'Mauritius', 'Mexico', 'Mongolia', 'Montenegro', 'Netherlands',
       'New Zealand', 'Nicaragua', 'Norway', 'Oman', 'Panama', 'Paraguay',
       'Philippines', 'Poland', 'Portugal', 'Puerto Rico', 'Qatar',
       'Republic of Korea', 'Romania', '

In [ ]:
df.country.nunique()

114

In [ ]:
df.country[:100]#All until 100

0     Albania
1     Albania
2     Albania
3     Albania
4     Albania
       ...   
95    Albania
96    Albania
97    Albania
98    Albania
99    Albania
Name: country, Length: 100, dtype: object

In [ ]:
#suicide_by_country = df["country"].value_counts().to_frame()
#high_number_of_data_points= suicide_by_country[suicide_by_country["country"] >= 200]
#low_number_of_data_points = suicide_by_country[suicide_by_country["country" ]<=199]
#suicide_by_country.reset_index(inplace=True)
#suicide_by_country.columns= ["country", "Counts"]
#suicide_by_country
#merged = df.merge(suicide_by_country, on="country", how="right")
#high_number_of_data_points = filtered_merged = merged[merged["Counts"] >= 200]
#high_number_of_data_points.groupby("country")["suicides_no"].sum().reset_index()
#sorted= high_number_of_data_points.sort_values(by= "suicides_no", ascending=False)
#suicide_cases_1987= sorted[sorted["year"] == 1987]
#suicide_cases_1987[:5]

country  year   sex          age  suicides_no  population  \
5875           Japan  1987  male  35-54 years       6688.0    18218000   
12173  United States  1987  male  35-54 years       6569.0    28222000   
12171  United States  1987  male  55-74 years       5482.0    18190000   
12172  United States  1987  male  25-34 years       5370.0    21670000   
12174  United States  1987  male  15-24 years       4109.0    19327000   

       suicides/100k_pop       country_year  HDI_for_year   gdp_for_year_($)_  \
5875               36.71          Japan1987           NaN  25,32,80,85,73,157   
12173              23.28  United States1987           NaN  48,70,21,70,00,000   
12171              30.14  United States1987           NaN  48,70,21,70,00,000   
12172              24.78  United States1987           NaN  48,70,21,70,00,000   
12174              21.26  United States1987           NaN  48,70,21,70,00,000   

       gdp_per_capita_($)       generation  Counts  
5875              22137.0           Silent     408  
12173             21631.0           Silent     372  
12171             21631.0  G.I. Generation     372  
12172             21631.0          Boomers     372  
12174             21631.0     Generation X     372

In [ ]:
suicides_per_100k= df.groupby("country")["suicides/100k_pop"].sum().reset_index().sort_values(by="suicides/100k_pop", ascending=False)
suicides_per_100k[:10]

country  suicides/100k_pop
82    Republic of Korea       19265.354105
85   Russian Federation       11327.335292
59            Lithuania       10685.115617
45              Hungary       10205.526298
53           Kazakhstan        9572.178274
6               Austria        9128.374664
107             Ukraine        8972.677881
51                Japan        8056.612843
37              Finland        7967.265858
12              Belgium        7931.355458

#Year


In [ ]:
df.year.unique()

array([1987, 1988, 1989, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
       2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
       1985, 1986, 1990, 1991, 2012, 2013, 2014, 2015, 2011, 2016, 2020,
       2019, 2018, 2017])

In [ ]:
#Group the years into 1980s, 1990s, 2000s, 2010s
df['decades'] = (10 * (df['year'] // 10)).astype(str) + 's'
df.head()

country  year     sex          age  suicides_no  population  \
0  Albania  1987    male  15-24 years         21.0      312900   
1  Albania  1987    male  35-54 years         16.0      308000   
2  Albania  1987  female  15-24 years         14.0      289700   
3  Albania  1987    male    75+ years          1.0       21800   
4  Albania  1987    male  25-34 years          9.0      274300   

   suicides/100k_pop country_year  HDI_for_year gdp_for_year_($)_  \
0               6.71  Albania1987           NaN    2,15,66,24,900   
1               5.19  Albania1987           NaN    2,15,66,24,900   
2               4.83  Albania1987           NaN    2,15,66,24,900   
3               4.59  Albania1987           NaN    2,15,66,24,900   
4               3.28  Albania1987           NaN    2,15,66,24,900   

   gdp_per_capita_($)       generation decades  
0               796.0     Generation X   1980s  
1               796.0           Silent   1980s  
2               796.0     Generation X   1980s  
3               796.0  G.I. Generation   1980s  
4               796.0          Boomers   1980s

In [ ]:
suicide_number_by_decades =df.groupby(["country", "decades"])["suicides/100k_pop"].sum().reset_index()
suicide_number_by_decades[suicide_number_by_decades["decades"] == "1980s"].sort_values(by="suicides/100k_pop", ascending=False)[:10]

country decades  suicides/100k_pop
23      Austria   1980s            1909.45
300   Singapore   1980s            1665.83
120      France   1980s            1656.95
45      Belgium   1980s            1543.14
59     Bulgaria   1980s            1471.68
173       Japan   1980s            1467.45
323    Suriname   1980s            1356.37
200  Luxembourg   1980s            1146.50
320   Sri Lanka   1980s            1088.52
116     Finland   1980s            1074.75

In [ ]:
suicide_number_by_decades[suicide_number_by_decades["decades"] == "1990s"].sort_values(by="suicides/100k_pop", ascending=False)[:10]

country decades  suicides/100k_pop
278  Russian Federation   1990s            4889.37
149             Hungary   1990s            4587.61
178          Kazakhstan   1990s            3956.98
42              Belarus   1990s            3867.16
350             Ukraine   1990s            3615.47
24              Austria   1990s            3300.49
117             Finland   1990s            3296.41
196           Lithuania   1990s            2928.24
121              France   1990s            2844.59
60             Bulgaria   1990s            2808.45

In [ ]:
suicide_number_by_decades[suicide_number_by_decades["decades"] == "2000s"].sort_values(by="suicides/100k_pop", ascending=False)[:10]

country decades  suicides/100k_pop
197           Lithuania   2000s            4890.56
279  Russian Federation   2000s            4342.43
270   Republic of Korea   2000s            4144.83
179          Kazakhstan   2000s            3885.17
150             Hungary   2000s            3632.19
192              Latvia   2000s            3422.66
309            Slovenia   2000s            3365.39
147              Guyana   2000s            3232.94
351             Ukraine   2000s            3128.35
111             Estonia   2000s            2989.67

In [ ]:
suicide_number_by_decades[suicide_number_by_decades["decades"] == "2010s"].sort_values(by="suicides/100k_pop", ascending=False)[:10]


country decades  suicides/100k_pop
271   Republic of Korea   2010s       12594.454105
307            Slovakia   2010s        3361.265754
198           Lithuania   2010s        2844.561592
151             Hungary   2010s        1985.726298
326            Suriname   2010s        1876.860000
180          Kazakhstan   2010s        1718.597062
280  Russian Federation   2010s        1696.715292
310            Slovenia   2010s        1670.871385
84              Croatia   2010s        1607.357897
193              Latvia   2010s        1606.978183

In [ ]:
Lithuania_suicide_rates= suicide_number_by_decades[suicide_number_by_decades["decades"] == "1980s"].sort_values(by="suicides/100k_pop", ascending=False)[:10]

In [ ]:
.groupby("decades")["population"].sum().reset_index()

decades  population
0   1990s    16811010
1   2000s    31819462
2   2010s   120543734
3   2020s    33538620

In [ ]:
result = df[df["country"] == "Lithuania"].groupby('decades').agg({'population': 'sum', 'suicides/100k_pop': 'sum'})
result

population  suicides/100k_pop
decades                               
1990s      16811010        2928.240000
2000s      31819462        4890.560000
2010s     120543734        2844.561592
2020s      33538620          21.754026

In [ ]:
result = df[df["country"] == "Lithuania"].groupby('decades').agg({'gdp_per_capita_($)': 'sum', 'suicides/100k_pop': 'sum'})
result

gdp_per_capita_($)  suicides/100k_pop
decades                                       
1990s          1.735440e+05        2928.240000
2000s          9.985440e+05        4890.560000
2010s          1.926571e+06        2844.561592
2020s          2.428037e+05          21.754026

In [ ]:
suicide_number_by_decades[suicide_number_by_decades["decades"] == "2010s"].sort_values(by="suicides/100k_pop", ascending=False)[:10]

country decades  suicides/100k_pop
271   Republic of Korea   2010s       12594.454105
307            Slovakia   2010s        3361.265754
198           Lithuania   2010s        2844.561592
151             Hungary   2010s        1985.726298
326            Suriname   2010s        1876.860000
180          Kazakhstan   2010s        1718.597062
280  Russian Federation   2010s        1696.715292
310            Slovenia   2010s        1670.871385
84              Croatia   2010s        1607.357897
193              Latvia   2010s        1606.978183

###Continent

In [ ]:
import pycountry_convert as pc

# Function to convert country to continent
def convert_country_to_continent(country):
    try:
        country_code = pc.country_name_to_country_alpha2(country)
        continent_code = pc.country_alpha2_to_continent_code(country_code)
        return pc.convert_continent_code_to_continent_name(continent_code)
    except:
        return 'Not found'

# Apply the function to the DataFrame
df['continent'] = df['country'].apply(convert_country_to_continent)

# Show the resulting DataFrame
df.head(5)

In [ ]:
# Group 'col2' values by 'col1' and get the number of unique values in each group
grouped = df.groupby('continent')['country'].nunique()
grouped
# Reset the index to make 'col1' and 'col2' columns
grouped = grouped.reset_index(name='count')
grouped.sort_values(by="count", ascending=False)

# Show the resulting DataFrame

#Questions and Insights to understand
1. How many countries were considered in this data? This data considered 114 countries; With top 3 most represented continents being from Europe, Asia and North America.
2. Which countries have the most data points? 82 countries have over 200 data points
2. Compare the top countries with the highest suicide numbers as per 100,000 people.  Republic of Korea, Russian Federation and Lithuania have the highest number of suicides per 100,0000 people cummulatively over the decades.
3. What is the trend of suicide numbers per 100,000 people over the decades?
In 1980s, Austria, Singapore and France had the highest number of suicide cases; In 1990s, Russian Federation had the highest number of suicide cases. In 2000s, Luthania doubled its number of suicide cases.* Need to investigate why there was an increase in suicide despite a significant increase in gdp per year.
4. Which age set is more likely to commit suicide in the Lithuania?
5. Which sex is more likey to commit suicide in Lithuania?
6. Which generation is more likely to commit suicide in Lithuania?

#Additional Things to query:
1. Why do we have 61% of missing values of HDI for year?